websocket/server.go

```diff
@@ -275,42 +275,45 @@ func (s *server) actuateRegisterClient(ctx context.Context, c *Client) {
 	// We judiciously ue the goroutines below, since are already on a goroutine!
 	<-readyChan
 	go s.broadcastMessageToTopic(ctx, MessageForBroadcast{
-		ToUserID: c.UserID,
+		ToUserID: c.UserID(),
 		Message: Message{
 			SessionID: c.SessionID(),
 			Topic: TopicDescriptor{
 				Type: TopicTypeAuth,
 			},
 		},
 	})
-	for _, topic := range c.Topics {
-		fmt.Println("NotifyTopicOfJoin", time.Now().Format("2006-01-02 15:04:05 Z"), c.UserID, topic)
+	for _, topic := range c.Topics() {
+		logrus.WithFields(logrus.Fields{
+			"topic":  topic,
+			"userID": c.UserID(),
+		}).Info("NotifyTopicOfJoin")
 		go s.broadcastMessageToTopic(ctx, MessageForBroadcast{
 			Message: Message{
 				FromUserID: SystemUserID,
 				Topic:      topic,
-				Text:       fmt.Sprintf("UserID '%s' has joined Room '%s'.", c.UserID, topic.ChatRoomName),
+				Text:       fmt.Sprintf("UserID '%s' has joined Room '%s'.", c.UserID(), topic.ChatRoomName),
 			},
 		})
 	}
 }
 
 // actuateRegisterClient is what the Server uses to actually de-register a Client, for whatever reason.
 // NOTE: This may be called on a goroutine, as it acts in reaction to data sent into a channel.
-func (s *server) actuateUnregisterClient(c *Client, err error) {
+func (s *server) actuateUnregisterClient(c Client, err error) {
 	logrus.WithFields(logrus.Fields{
-		"c.UserID": c.UserID,
+		"c.UserID": c.UserID(),
 	}).Infof("UnregisterClient")
 
 	// Unregister Client from all Topics, then remove Client.
 	s.mutex.Lock()
 	defer s.mutex.Unlock()
-	if _, ok := s.clientsByUserID[c.UserID]; ok {
-		delete(s.clientsByUserID, c.UserID)
-		close(c.send)
```

Comment:
Removed this from this call and added to the `client.Close(error)` that is called at the end of this method. It should be the responsibility of the client to close its internal structures.